src/topics/scheduled.js

```diff
@@ -0,0 +1,108 @@
+'use strict';
+
+const winston = require('winston');
+const { CronJob } = require('cron');
+
+const db = require('../database');
+const categories = require('../categories');
+const posts = require('../posts');
+const socketHelpers = require('../socket.io/helpers');
+const topics = require('./index');
+const user = require('../user');
+
+const Scheduled = module.exports;
+
+Scheduled.startJobs = function () {
+	winston.verbose('[scheduled topics] Starting jobs.');
+	new CronJob('*/1 * * * *', Scheduled.handleExpired, null, true);
+};
+
+Scheduled.handleExpired = async function () {
+	const now = Date.now();
+	const tids = await db.getSortedSetRangeByScore('topics:scheduled', 0, -1, '-inf', now);
+
+	if (!tids.length) {
+		return;
+	}
+
+	let topicsData = await topics.getTopicsData(tids);
+	// Filter deleted
+	topicsData = topicsData.filter(topicData => Boolean(topicData));
+	const uids = topicsData
+		.map(topicData => topicData.uid)
+		// Filter guests topics
+		.filter(uid => uid)
+		.filter((uid, idx, arr) => arr.indexOf(uid) === idx);
+
+	await Promise.all([].concat(
+		sendNotifications(uids, topicsData),
+		updateUserLastposttimes(uids, topicsData),
+		topicsData.map(topicData => categories.updateRecentTid(topicData.cid, topicData.tid)),
+		topicsData.map(topicData => unpin(topicData.tid, topicData)),
+		updateUnreadCounts(),
+		db.sortedSetsRemoveRangeByScore([`topics:scheduled`], '-inf', now)
+	));
+};
+
+Scheduled.pin = async function (tid, topicData) {
+	return [
```

Comment:
You can use topics.tools.pin instead

Reply:
The privilege check in topics/tools.js#togglePin would block non-admin/mod users with "topics:schedule" privilege. But since we're going to limit scheduling to admin/mods now, can remove these.

Reply:
Just as a historical note, not going with the limiting mentioned above, so these stay.